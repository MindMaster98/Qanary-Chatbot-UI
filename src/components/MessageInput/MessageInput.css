/* input box styles go here */

#messageInput {
  position: relative;
  height: 150px;
  width: 100%;
  border-top: 2px solid #f0f0f0;
  display: flex;
  flex-wrap: nowrap;
  align-items: stretch;
}

#inputTextBox {
  position: relative;
  font-size: x-large;
  width: 90%;
  resize: none;
  display: inline-block;
}

#inputTextBox,
#inputTextBox:focus {
  border: none;
  overflow: auto;
  outline: none;
  -webkit-box-shadow: none;
  -moz-box-shadow: none;
  box-shadow: none;
}

#submitButtonContainer {
  position: relative;
  width: 200px;
  display: inline-block;
  flex-shrink: 0;
}

/* loading bar styles go here */

#loadingBar {
  background-color: gray;
}

#loadingBar.visible {
  height: 15px;
  border-radius: 7px;
}

@keyframes load {
  0% {
    left: 0%;
  }
  100% {
    left: 80%;
  }
}

#loadingBar.visible::before {
  content: '';
  position: relative;
  display: block;
  left: 0%;
  top: 2.5px;
  width: 20%;
  border-radius: 7px;
  height: 7.5px;
  background-color: #aa2941;
  animation: load 1.75s infinite;
}