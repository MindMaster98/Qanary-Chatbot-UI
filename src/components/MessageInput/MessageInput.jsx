import React, { Component } from "react";
import { Button } from "..";
import "./MessageInput.css";
import texts from "../../texts/de/texts.json";

export default class MessageInput extends Component {
  constructor(props) {
    super(props);

    this.state = {
      isSending: false,
      query: "",
    };

    this.onSendMessage = this.onSendMessage.bind(this);
    this.inputTextChange = this.inputTextChange.bind(this);
    this.inputKeyPress = this.inputKeyPress.bind(this);
  }

  onSendMessage() {
    this.props.sendMessage(this.state.query);
    this.setState({ isSending: true, query: "" });
  }

  inputTextChange(changeEvent) {
    this.setState({ query: changeEvent.target.value });
  }

  inputKeyPress(keyPressEvent) {
    if (keyPressEvent.key === "Enter") {
      this.onSendMessage();
    }
  }

  render() {
    return (
      <>
        <div
          id="loadingBar"
          className={this.state.isSending ? "visible" : ""}
        />
        <div id="messageInput">
          <textarea
            id="inputTextBox"
            placeholder={texts["message-input"]["placeholder-text"]}
            value={this.state.query}
            onChange={this.inputTextChange}
            onKeyPress={this.inputKeyPress}
          />
          <div id="submitButtonContainer">
            <Button
              text={texts["message-input"]["submit-button"]}
              onClick={this.onSendMessage}
              style={{
                width: "150px",
                height: "50px",
                marginRight: "30px",
              }}
              className={"light-pink vertical-center right-side"}
            />
          </div>
        </div>
      </>
    );
  }
}
