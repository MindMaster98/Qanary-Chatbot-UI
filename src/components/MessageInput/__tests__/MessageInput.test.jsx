import renderer from "react-test-renderer";
import { act } from "@testing-library/react";
import userEvent from '@testing-library/user-event'
import ReactDOM from "react-dom";
import MessageInput from "../MessageInput";

// Snapshot tests

it("renders correctly with no props", () => {
  const tree = renderer.create(<MessageInput />).toJSON();
  expect(tree).toMatchSnapshot();
});

it("renders correctly with loading bar", () => {
  const tree = renderer.create(<MessageInput isSending={true} />).toJSON();
  expect(tree).toMatchSnapshot();
});

it("renders correctly without loading bar", () => {
  const tree = renderer.create(<MessageInput isSending={false} />).toJSON();
  expect(tree).toMatchSnapshot();
});

// Blackbox tests
// You cannot dispatch events using test renderer, so I need to actually add the element to the DOM
// and dispatch an actual event. Calling the function directly would oppose the idea of a blackbox.
// Why not enzyme? Because enzyme doesn't officially support React 17.

let container;

beforeEach(() => {
  container = document.createElement("div");
  document.body.appendChild(container);
});

afterEach(() => {
  document.body.removeChild(container);
  container = null;
});

it("sendMessage function is called on click with valid query", async () => {
  const mockFn = jest.fn();

  act(() => {
    ReactDOM.render(<MessageInput sendMessage={mockFn} />, container);
  });

  const textarea = document.querySelector("#messageInput > textarea");
  const button = document.querySelector("#submitButtonContainer > button");
  userEvent.type(textarea, "test");
  
  button.dispatchEvent(new InputEvent("click", { bubbles: true }));

  expect(mockFn).toHaveBeenCalledWith("test");
});

it("sendMessage function isn't called on click with empty query", () => {
  const mockFn = jest.fn();

  act(() => {
    ReactDOM.render(<MessageInput sendMessage={mockFn} />, container);
  });

  const button = document.querySelector("#submitButtonContainer > button");
  button.click();

  expect(mockFn).toBeCalledTimes(0);
});
