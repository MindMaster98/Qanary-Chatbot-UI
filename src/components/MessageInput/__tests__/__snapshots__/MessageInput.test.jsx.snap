// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with loading bar 1`] = `
Array [
  <div
    className="visible"
    id="loadingBar"
    tabIndex="-1"
  />,
  <div
    id="messageInput"
    tabIndex="-1"
  >
    <textarea
      className="black-text"
      id="inputTextBox"
      onChange={[Function]}
      onKeyPress={[Function]}
      placeholder="Ihre Frage an mich..."
      value=""
    />
    <div
      id="submitButtonContainer"
      tabIndex="-1"
    >
      <button
        className="button light-gray black-text vertical-center right-side"
        onClick={[Function]}
        style={
          Object {
            "height": "40px",
            "marginRight": "10px",
            "width": "100px",
          }
        }
      >
        Senden
      </button>
    </div>
  </div>,
]
`;

exports[`renders correctly with no props 1`] = `
Array [
  <div
    className=""
    id="loadingBar"
    tabIndex="-1"
  />,
  <div
    id="messageInput"
    tabIndex="-1"
  >
    <textarea
      className="black-text"
      id="inputTextBox"
      onChange={[Function]}
      onKeyPress={[Function]}
      placeholder="Ihre Frage an mich..."
      value=""
    />
    <div
      id="submitButtonContainer"
      tabIndex="-1"
    >
      <button
        className="button light-gray black-text vertical-center right-side"
        onClick={[Function]}
        style={
          Object {
            "height": "40px",
            "marginRight": "10px",
            "width": "100px",
          }
        }
      >
        Senden
      </button>
    </div>
  </div>,
]
`;

exports[`renders correctly without loading bar 1`] = `
Array [
  <div
    className=""
    id="loadingBar"
    tabIndex="-1"
  />,
  <div
    id="messageInput"
    tabIndex="-1"
  >
    <textarea
      className="black-text"
      id="inputTextBox"
      onChange={[Function]}
      onKeyPress={[Function]}
      placeholder="Ihre Frage an mich..."
      value=""
    />
    <div
      id="submitButtonContainer"
      tabIndex="-1"
    >
      <button
        className="button light-gray black-text vertical-center right-side"
        onClick={[Function]}
        style={
          Object {
            "height": "40px",
            "marginRight": "10px",
            "width": "100px",
          }
        }
      >
        Senden
      </button>
    </div>
  </div>,
]
`;
